File: src/NFramework.NFW/core/NFramework.NFW.Application/Features/TemplateManagement/Queries/ListTemplates/ListTemplatesQueryHandler.cs
using NFramework.NFW.Application.Features.Cli.Logging;
using NFramework.NFW.Application.Features.TemplateManagement.Services;
using NFramework.NFW.Application.Features.Versioning.Abstractions;
using NFramework.NFW.Domain.Features.TemplateManagement.ValueObjects;
using NFramework.NFW.Domain.Features.Version;

namespace NFramework.NFW.Application.Features.TemplateManagement.Queries.ListTemplates;

public sealed class ListTemplatesQueryHandler(
    TemplatesService templatesService,
    IVersionProvider versionProvider,
    DiagnosticLogger diagnosticLogger
)
{
    private readonly TemplatesService _templatesService = templatesService;
    private readonly IVersionProvider _versionProvider = versionProvider;
    private readonly DiagnosticLogger _diagnosticLogger = diagnosticLogger;

    public async Task<ListTemplatesQueryResult> HandleAsync(
        ListTemplatesQuery query,
        CancellationToken cancellationToken
    )
    {
        ArgumentNullException.ThrowIfNull(query);

        try
        {
            _diagnosticLogger.Write("Resolving template catalog source.");
            VersionInfo versionInfo = _versionProvider.GetVersionInfo();
            TemplateCatalog catalog = await _templatesService.GetCatalogAsync(
                versionInfo.SemanticVersion,
                cancellationToken
            );
            ListedTemplate[] templates = catalog
                .Entries.Select(template => new ListedTemplate(
                    template.Identifier,
                    template.DisplayName,
                    template.Description
                ))
                .ToArray();

            return ListTemplatesQueryResult.Success(templates);
        }
        catch (TemplateCatalogException exception)
        {
            _diagnosticLogger.Write(exception.ToString());
            return ListTemplatesQueryResult.FailureResult(exception.Message);
        }
    }
}

File: src/nframework-nfw/core/nframework-nfw-application/src/features/template_management/queries/list_templates/list_templates_query_handler.rs
use crate::features::template_management::models::errors::templates_service_error::TemplatesServiceError;
use crate::features::template_management::queries::list_templates::list_templates_query::ListTemplatesQuery;
use crate::features::template_management::queries::list_templates::list_templates_query_result::ListTemplatesQueryResult;
use crate::features::template_management::services::abstraction::template_listing_service::TemplateListingService;

#[derive(Debug, Clone)]
pub struct ListTemplatesQueryHandler<S>
where
    S: TemplateListingService,
{
    templates_service: S,
}

impl<S> ListTemplatesQueryHandler<S>
where
    S: TemplateListingService,
{
    pub fn new(templates_service: S) -> Self {
        Self { templates_service }
    }

    pub fn handle(
        &self,
        _query: ListTemplatesQuery,
    ) -> Result<ListTemplatesQueryResult, TemplatesServiceError> {
        let (templates, warnings) = self.templates_service.list_templates()?;
        Ok(ListTemplatesQueryResult {
            templates,
            warnings,
        })
    }
}
